File: spring-reactive-01/learn-java/src/main/java/boki/learnjava/backpressure/strategy/DropEx.java
package boki.learnjava.backpressure.strategy;

import boki.learnjava.util.Logger;
import boki.learnjava.util.TimeUtils;
import reactor.core.publisher.Flux;
import reactor.core.scheduler.Schedulers;
import java.time.Duration;

/**
 * Unbounded request 일 경우, Downstream 에 Backpressure Drop 전략을 적용하는 예제
 *  - Downstream 으로 전달 할 데이터가 버퍼에 가득 찰 경우, 버퍼 밖에서 대기하는 먼저 emit 된 데이터를 Drop 시키는 전략
 */
public class DropEx {
    public static void main(String[] args) {
        Flux
            .interval(Duration.ofMillis(1L))
            .onBackpressureDrop(dropped -> Logger.info("# dropped: {}", dropped))
            .publishOn(Schedulers.parallel())
            .subscribe(data -> {
                    TimeUtils.sleep(5L);
                    Logger.onNext(data);
                },
                Logger::onError);

        TimeUtils.sleep(2000L);
    }
}

File: spring-reactive-01/learn-kt/src/main/kotlin/boki/learn/backpressure/strategy/DropEx.kt
package boki.learn.backpressure.strategy

import reactor.core.publisher.Flux
import reactor.core.scheduler.Schedulers
import java.time.Duration

/**
 * Unbounded request 일 경우, Downstream 에 Backpressure Drop 전략을 적용하는 예제
 *  - Downstream 으로 전달 할 데이터가 버퍼에 가득 찰 경우, 버퍼 밖에서 대기하는 먼저 emit 된 데이터를 Drop 시키는 전략
 */
fun main() {
    Flux
        .interval(Duration.ofMillis(1L))
        .onBackpressureDrop { dropped -> println("# dropped: $dropped") }
        .publishOn(Schedulers.parallel())
        .subscribe(
            { data: Long ->
                Thread.sleep(5L)
                println("onNext $data")
            },
            { error: Throwable -> println("onError $error") })

    Thread.sleep(2000L)
}